### DISCUSSION TITLE
Use `cargo check --workspace`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1948] [Announcements]
Over the last few months I have lamented that `cargo check` did not check `fuzz` even though we recently included it in the workspace - turns out the problem was between the keyboard and the chair. The correct invocation to `cargo` should include `--workspace` (or `--all`).

Same for other crates with fuzzing (`rust-secp`, `rust-miniscript` etc).

**Comment by apoelstra** [comment]
Definitely ACK, but what a frustrating situation ... we have `--all-features`, `--all-targets`, `--all`, `--workspace` ... and only the first one do I actually know what it does. And I'm not sure which cargo commands have which flags.

**Reply by tcharding** [reply]
All right, I did not know the answer either so I took one for the team and worked this out.

`--all` is a deprecated alias for `--workspace`, I'll use `--workspace` because its more clear.

## For cargo clippy/check/test

1. `--all-features` is needed to include feature gated code
2. `--all-targets` is needed to include test/bench code as well as `examples/` code
3. `--workspace` is needed to include all the crates in the workspace

## For cargo fmt

We get 1 and 2 above for free, need `--all` to get workspace (fuzz) code included. 

## Local dev

I'm going to create shell aliases to:

- `cargo check/clippy/test --workspace --all-features --all-targets`
- `cargo fmt +nightly --all`

## CI / githooks

I'll PR to see that we are checking the things we want, I've had a few thoughts recently that lead me to think the CI scripts need some attention again.

**Reply by tcharding** [reply]
Phew!

**Reply by tcharding** [reply]
In case its useful:
```bash
CARGO_FLAGS=' --color=always --workspace --all-features --all-targets'
alias cc="cargo check $CARGO_FLAGS"
alias cb="cargo build $CARGO_FLAGS"
alias ct="cargo test $CARGO_FLAGS"
alias cf="cargo +nightly fmt --all"
```

**Comment by Shong1111** [comment]
# - [ ] _~~~~_

**Comment by tcharding** [comment]
Whats the go with all the spam posts on github right now?

**Reply by Shong1111** [reply]
What you think 🤔

**Reply by tcharding** [reply]
Oh you are a real person, sorry bud. I don't know what `- [ ] ~~~~` means?
